src/H5TSint.c

```diff
@@ -0,0 +1,577 @@
+        HGOTO_DONE(NULL);
+    if (NULL == tinfo_node)
+        /* Create thread info for this thread */
+        if (H5_UNLIKELY(NULL == (tinfo_node = H5TS__tinfo_create())))
+            HGOTO_DONE(NULL);
+
+    /* Set return value */
+    ret_value = &tinfo_node->info.api_ctx_node_ptr;
+
+done:
+    FUNC_LEAVE_NOAPI_NAMECHECK_ONLY(ret_value)
+} /* H5TS_get_api_ctx_ptr() */
+
+/*--------------------------------------------------------------------------
+ * Function:    H5TS_get_err_stack
+ *
+ * Purpose:     Retrieve the address of error stack for this thread.
+ *		(i.e. an H5E_stack_t *)
+ *
+ * Return:	Success: Non-NULL pointer to error stack for thread
+ *		Failure: NULL
+ *
+ *--------------------------------------------------------------------------
+ */
+H5E_stack_t *
+H5TS_get_err_stack(void)
+{
+    H5TS_tinfo_node_t *tinfo_node;
+    H5E_stack_t       *ret_value;
+
+    FUNC_ENTER_NOAPI_NAMECHECK_ONLY
+
+    /* Check if info for thread has been created */
+    if (H5_UNLIKELY(H5TS_key_get_value(H5TS_thrd_info_key_g, (void **)&tinfo_node) < 0))
+        HGOTO_DONE(NULL);
+    if (NULL == tinfo_node)
+        /* Create thread info for this thread */
+        if (H5_UNLIKELY(NULL == (tinfo_node = H5TS__tinfo_create())))
+            HGOTO_DONE(NULL);
+
+    /* Set return value */
+    ret_value = &tinfo_node->info.err_stack;
+
+done:
+    FUNC_LEAVE_NOAPI_NAMECHECK_ONLY(ret_value)
+} /* H5TS_get_err_stack() */
+
+/*--------------------------------------------------------------------------
+ * Function:    H5TS__tinfo_destroy
+ *
+ * Purpose:     When a thread shuts down, put its info record on the free list.
+ *
+ * Note:      	This routine runs asynchronously _outside_ of the library and
+ *		is not covered by the library's API lock.  Therefore, protect
+ *		access to the global variable with a mutex.
+ *
+ * Return:      None
+ *
+ *--------------------------------------------------------------------------
+ */
+void
+H5TS__tinfo_destroy(void *_tinfo_node)
+{
+    H5TS_tinfo_node_t *tinfo_node = _tinfo_node;
+
+    FUNC_ENTER_PACKAGE_NAMECHECK_ONLY
+
+    if (tinfo_node) {
+        /* Add thread info node to the free list */
+        H5TS_mutex_lock(&H5TS_tinfo_mtx_s);
+        tinfo_node->next       = H5TS_tinfo_next_free_s;
+        H5TS_tinfo_next_free_s = tinfo_node;
+        H5TS_mutex_unlock(&H5TS_tinfo_mtx_s);
+    }
+
+    FUNC_LEAVE_NOAPI_VOID_NAMECHECK_ONLY
+}
+
+/*--------------------------------------------------------------------------
+ * Function:    H5TS__tinfo_term
+ *
+ * Purpose:     Terminate per-thread info at library shutdown
+ *
+ * Return:      Non-negative on success / Negative on failure
+ *
+ *--------------------------------------------------------------------------
+ */
+herr_t
+H5TS__tinfo_term(void)
+{
+    herr_t ret_value = SUCCEED;
+
+    FUNC_ENTER_PACKAGE_NAMECHECK_ONLY
+
+    /* Release nodes on the free list */
+    if (H5_UNLIKELY(H5TS_mutex_lock(&H5TS_tinfo_mtx_s) < 0))
```

Comment:
If there's possible contention for this mutex here is it safe to destroy it later in this function?

Reply:
Yes, this should be fine since it is called from H5TS_term_package(), which is called from the atexit() handler for the process and will only be invoked from the main thread.  Technically, acquiring the mutex is unnecessary, but seemed reasonable.